tasks/unseal/task_unseal_decode.go

```diff
@@ -0,0 +1,251 @@
+		FROM sectors_unseal_pipeline
+		WHERE task_id_decode_sector = $1`, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("getting sector params: %w", err)
+	}
+
+	if len(sectorParamsArr) == 0 {
+		return false, xerrors.Errorf("no sector params")
+	}
+
+	sectorParams := sectorParamsArr[0]
+
+	var sectorMeta []struct {
+		TicketValue    []byte `db:"ticket_value"`
+		OrigSealedCID  string `db:"orig_sealed_cid"`
+		CurSealedCID   string `db:"cur_sealed_cid"`
+		CurUnsealedCID string `db:"cur_unsealed_cid"`
+	}
+	err = t.db.Select(ctx, &sectorMeta, `
+		SELECT ticket_value, orig_sealed_cid, cur_sealed_cid, cur_unsealed_cid
+		FROM sectors_meta
+		WHERE sp_id = $1 AND sector_num = $2`, sectorParams.SpID, sectorParams.SectorNumber)
+	if err != nil {
+		return false, xerrors.Errorf("getting sector meta: %w", err)
+	}
+
+	if len(sectorMeta) != 1 {
+		return false, xerrors.Errorf("expected 1 sector meta, got %d", len(sectorMeta))
+	}
+
+	smeta := sectorMeta[0]
+	commK, err := cid.Decode(smeta.OrigSealedCID)
+	if err != nil {
+		return false, xerrors.Errorf("decoding OrigSealedCID: %w", err)
+	}
+
+	var commD, commR cid.Cid
+	if smeta.CurSealedCID == "" || smeta.CurSealedCID == "b" {
+		// https://github.com/filecoin-project/curio/issues/191
+		// <workaround>
+
+		// "unsealed" actually stores the sealed CID, "sealed" is empty
+		commR, err = cid.Decode(smeta.CurUnsealedCID)
+		if err != nil {
+			return false, xerrors.Errorf("decoding CurSealedCID: %w", err)
+		}
+
+		commD, err = dealdata.UnsealedCidFromPieces(ctx, t.db, sectorParams.SpID, sectorParams.SectorNumber)
+		if err != nil {
+			return false, xerrors.Errorf("getting deal data CID: %w", err)
+		}
+
+		log.Warnw("workaround for issue #191", "task", taskID, "commD", commD, "commK", commK, "commR", commR)
+
+		// </workaround>
+	} else {
+		commD, err = cid.Decode(smeta.CurUnsealedCID)
+		if err != nil {
+			return false, xerrors.Errorf("decoding CurUnsealedCID (%s): %w", smeta.CurUnsealedCID, err)
+		}
+		commR, err = cid.Decode(smeta.CurSealedCID)
+		if err != nil {
+			return false, xerrors.Errorf("decoding CurSealedCID: %w", err)
+		}
+	}
+
+	sref := storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(sectorParams.SpID),
+			Number: abi.SectorNumber(sectorParams.SectorNumber),
+		},
+		ProofType: abi.RegisteredSealProof(sectorParams.RegSealProof),
+	}
+
+	isSnap := commK != commR
+	log.Infow("unseal decode", "snap", isSnap, "task", taskID, "commK", commK, "commR", commR, "commD", commD)
+	if isSnap {
+		err := t.sc.DecodeSnap(ctx, taskID, commD, commK, sref)
+		if err != nil {
+			return false, xerrors.Errorf("DecodeSnap: %w", err)
+		}
+	} else {
+		err = t.sc.DecodeSDR(ctx, taskID, sref)
+		if err != nil {
+			return false, xerrors.Errorf("DecodeSDR: %w", err)
+		}
+	}
+
+	_, err = t.db.Exec(ctx, `UPDATE sectors_unseal_pipeline SET after_decode_sector = TRUE, task_id_decode_sector = NULL WHERE task_id_decode_sector = $1`, taskID)
```

Comment:
Where do we delete the key file?

Reply:
Bottom of SealCalls.decodeCommon, will add a comment here